### PR TITLE
CMTS parsing and RIP

### PR DESCRIPTION
**Pull request by arifogel**
- Cisco CMTS parsing, including new structure types
- RIP full-pipeline implementation
- Various other parsing

### CODE REVIEW

**Comment by ratulm** [comment]
I didn't delete the branch in case you need to add to it. Feel free to delete when its no longer needed.

File: projects/batfish/src/main/java/org/batfish/bdp/BdpDataPlanePlugin.java [hunk 1]
```diff
@@ -415,7 +416,59 @@ private void computeFixedPoint(
...
+              "Compute RIP Internal routes: iteration " + ripInternalIterations, nodes.size());
+      nodes
+          .values()
+          .parallelStream()
+          .forEach(
+              n -> {
+                for (VirtualRouter vr : n._virtualRouters.values()) {
+                  if (vr.propagateRipInternalRoutes(nodes, topology)) {
+                    synchronized (routesChangedMonitor) {
+                      ripInternalChanged[0] = true;
```
Comment (dhalperi) [inline projects/batfish/src/main/java/org/batfish/bdp/BdpDataPlanePlugin.java]: any reason not to use an AtomicBoolean for this?

File: projects/batfish/src/main/java/org/batfish/grammar/cisco/CiscoControlPlaneExtractor.java [hunk 2]
```diff
@@ -4116,14 +4340,36 @@ public void exitRedistribute_ospf_bgp_tail(Redistribute_ospf_bgp_tailContext ctx
   }
 
   @Override
+  public void exitRedistribute_rip_bgp_tail(Redistribute_rip_bgp_tailContext ctx) {
+    BgpProcess proc = currentVrf().getBgpProcess();
+    if (_currentPeerGroup == proc.getMasterBgpPeerGroup()) {
```
Comment (dhalperi) [inline projects/batfish/src/main/java/org/batfish/grammar/cisco/CiscoControlPlaneExtractor.java]: `==` or `.equals`?

File: projects/batfish/src/main/java/org/batfish/grammar/cisco/CiscoControlPlaneExtractor.java [hunk 3]
```diff
@@ -5483,6 +5785,21 @@ private IntExpr toCommonIntExpr(Int_exprContext ctx) {
...
 
+  private LongExpr toCommonLongExpr(Int_exprContext ctx) {
+    if (ctx.DEC() != null && ctx.PLUS() == null && ctx.DASH() == null) {
+      long val = toLong(ctx.DEC());
+      return new LiteralLong(val);
+    } else if (ctx.RP_VARIABLE() != null) {
+      return new VarLong(ctx.RP_VARIABLE().getText());
+    } else {
+      /*
+       * Unsupported static integer expression - do not add cases unless you
```
Comment (dhalperi) [inline projects/batfish/src/main/java/org/batfish/grammar/cisco/CiscoControlPlaneExtractor.java]: long

File: projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/RipInternalRoute.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    super(network, nextHopIp, admin, metric);
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (this == obj) {
+      return true;
+    }
+    RipInternalRoute other = (RipInternalRoute) obj;
+    if (_nextHopIp == null) {
```
[the remark this answers is not in the document]
Comment (arifogel) [reply projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/RipInternalRoute.java]: Will address.

File: projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/RipInternalRoute.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  protected final String protocolRouteString() {
+    return "";
+  }
+
+  @Override
+  public int routeCompare(AbstractRoute rhs) {
+    if (getClass() != rhs.getClass()) {
+      return 0;
+    }
+    return 0;
```
[the remark this answers is not in the document]
Comment (arifogel) [reply projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/RipInternalRoute.java]: Extra fields removed when copied from OSPF. Now the function looks trivial. Will look over.

File: projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/vendor_family/cisco/DocsisPolicy.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
+package org.batfish.datamodel.vendor_family.cisco;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.util.ArrayList;
+import java.util.List;
+import org.batfish.common.util.ComparableStructure;
+
+public class DocsisPolicy extends ComparableStructure<String> {
```
[the remark this answers is not in the document]
Comment (arifogel) [reply projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/vendor_family/cisco/DocsisPolicy.java]: That's the whole point of ComparableStructure. It provides a base implementation of those, using the name of the structure as the sole field for comparison and hashing.

File: projects/batfish/src/main/java/org/batfish/bdp/BdpDataPlanePlugin.java (hunk 1 above)
Comment (arifogel) [reply projects/batfish/src/main/java/org/batfish/bdp/BdpDataPlanePlugin.java]: Didn't know about it at the time. Should do the trick.

File: projects/batfish/src/main/java/org/batfish/grammar/cisco/CiscoControlPlaneExtractor.java (hunk 2 above)
Comment (arifogel) [reply projects/batfish/src/main/java/org/batfish/grammar/cisco/CiscoControlPlaneExtractor.java]: ==. This is meant to be compare-by-reference. I will add comment.

File: projects/batfish/src/main/java/org/batfish/grammar/cisco/CiscoControlPlaneExtractor.java (hunk 3 above)
Comment (arifogel) [reply projects/batfish/src/main/java/org/batfish/grammar/cisco/CiscoControlPlaneExtractor.java]: Right.

File: projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/RipInternalRoute.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+      if (other._nextHopIp != null) {
+        return false;
+      }
+    } else if (!_nextHopIp.equals(other._nextHopIp)) {
+      return false;
+    }
+    if (_admin != other._admin) {
+      return false;
+    }
+    if (_metric != other._metric) {
```
[the remark this answers is not in the document]
Comment (arifogel) [reply projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/RipInternalRoute.java]: admin is unlikely to ever be different, but thrown in there for completeness. I don't want to hack all these right now, but I take your point.

**Pull request merged**